### ISSUE TITLE
DrawUnitIconsScreen: Unit icons often missing at oblique camera angles. More info here. 

### ISSUE THREAD

**Issue by Beherith** [beyond-all-reason/spring#1175]
I added some output to colorize the unit icons on failure cases of not being drawn. 

Icons are colored white then they fail the

`if (modelDrawerData->iconZoomDist / unitRadiusMult < modelDrawerData->iconFadeVanish)`

Check. 

Below shot demonstrates that all the units failed this check (notwithstanding those that also failed the Z checks in the air)

![image](https://github.com/beyond-all-reason/spring/assets/109391/e2d08354-a56d-450c-8cf0-ba7a4a9aedf5)

`

		for (const CUnit* unit : units)
		{

			SColor color = unit->isSelected ? color4::white : SColor{ teamHandler.Team(unit->team)->color };
			if (!unit->drawIcon)
				color = color4::red;
				//continue;

			const bool canSee = gu->spectatingFullView || (unit->losStatus[gu->myAllyTeam] && (LOS_INLOS | LOS_CONTRADAR | LOS_PREVLOS) == (LOS_INLOS | LOS_CONTRADAR | LOS_PREVLOS));
			if (!canSee)
				color = color4::yellow;
				//continue;

			assert(unit->myIcon == icon);
			// iconUnits should not never contain void-space units, see UpdateUnitIconState
			assert(!unit->IsInVoid());

			// drawMidPos is auto-calculated now; can wobble on its own as pieces move
			float3 pos = (!gu->spectatingFullView) ?
				unit->GetObjDrawErrorPos(gu->myAllyTeam) :
				unit->GetObjDrawMidPos();

			pos = camera->CalcViewPortCoordinates(pos);
			if (pos.z > 1.0f || pos.z < 0.0f)
				color = color4::green;
				//continue;

			// use white for selected units

			float unitRadiusMult = icon->GetSize();
			if (icon->GetRadiusAdjust() && icon != icon::iconHandler.GetDefaultIconData())
				unitRadiusMult *= (unit->radius / icon->GetRadiusScale());
			unitRadiusMult = (unitRadiusMult - 1) * 0.75 + 1;

			// fade icons away in high zoom in levels
			if (!unit->GetIsIcon()) {
				if (modelDrawerData->iconZoomDist / unitRadiusMult < modelDrawerData->iconFadeVanish)
					//continue;
					color = color4::white;
				else if (modelDrawerData->iconFadeVanish < modelDrawerData->iconFadeStart && modelDrawerData->iconZoomDist / unitRadiusMult < modelDrawerData->iconFadeStart)
					// alpha range [64, 255], since icons is unrecognisable with alpha < 64
					color.a = 64 + 191.0f * (modelDrawerData->iconZoomDist / unitRadiusMult - modelDrawerData->iconFadeVanish) / (modelDrawerData->iconFadeStart - modelDrawerData->iconFadeVanish);
			}
`

Based

**Comment by Beherith** [comment]
This part looks completely broken, why is it applied on a blanket all-icons level?

If it is, then why is it calculated for every unit? Does GCC optimize this stuff out of the loop?
``
			// fade icons away in high zoom in levels
			if (!unit->GetIsIcon()) {
				if (modelDrawerData->iconZoomDist / unitRadiusMult < modelDrawerData->iconFadeVanish)
					continue;
``

**Comment by Beherith** [comment]
It turns out that `modelDrawerData->iconFadeVanish` is the value that is incorrectly calculated at oblique angles.

**Comment by lhog** [comment]
Yes, this part, like every other one using trace rays tends to fail at oblique angles or weird zoom levels.

**Comment by lhog** [comment]
Let's see if this can be verified too: https://github.com/ZeroK-RTS/Zero-K/issues/4809#issuecomment-1876937766
